webapp/channels/src/components/admin_console/access_control/editors/cel_editor/editor.tsx

```diff
@@ -0,0 +1,363 @@
+                isValidating: false,
+            }));
+            onValidate?.(isValid);
+        } catch (error) {
+            setEditorState((prev) => ({
+                ...prev,
+                isValid: false,
+                validationErrors: [error.detailed_error || 'Unknown error'],
+                statusBarColor: 'var(--error-text)',
+                isValidating: false,
+            }));
+            onValidate?.(false);
+        }
+    }, [editorState.expression, onValidate]);
+
+    // initialize monaco editor
+    useEffect(() => {
+        if (!editorRef.current || monacoRef.current) {
+            return () => {};
+        }
+
+        monacoRef.current = monaco.editor.create(editorRef.current, MONACO_EDITOR_OPTIONS);
+
+        // Set the initial value from the expression state
+        monacoRef.current.setValue(editorState.expression);
+
+        monacoRef.current.getModel()?.onDidChangeContent(() => {
+            const newValue = monacoRef.current?.getValue() || '';
+            handleChange(newValue);
+        });
+
+        monacoRef.current.onDidChangeCursorPosition((e) => {
+            setEditorState((prev) => ({
+                ...prev,
+                cursorPosition: {line: e.position.lineNumber, column: e.position.column},
+            }));
+        });
+
+        // To disable monaco's default behavior of opening the find and replace widget
+        monaco.editor.addKeybindingRule({
+            keybinding: monaco.KeyMod.CtrlCmd | monaco.KeyCode.KeyF,
+            command: null,
+        });
+
+        monaco.editor.addCommand({
+            id: VALIDATE_POLICY_SYNTAX_COMMAND_ID,
+            run: validateSyntax,
+        });
+
+        monaco.editor.addKeybindingRule({
+            keybinding: monaco.KeyMod.Alt | monaco.KeyCode.Enter,
+            command: VALIDATE_POLICY_SYNTAX_COMMAND_ID,
+        });
+
+        return () => {
+            if (monacoRef.current) {
+                monacoRef.current.dispose();
+                monacoRef.current = null;
+            }
+        };
+    }, []);
+
+    return (
+        <div className={`cel-editor ${className}`}>
+            <MonacoLanguageProvider schemas={schemas}/>
+
+            <div
+                className='cel-editor__container'
+                data-status-color={editorState.statusBarColor}
+            >
+                {!editorState.expression && (
+                    <div
+                        className='policy-editor-placeholder'
+                        aria-label='CEL Expression Editor'
+                    >
+                        {placeholder}
+                    </div>
+                )}
+
+                <div
+                    ref={editorRef}
+                    className='cel-editor__input'
+                />
+                <div
+                    className='cel-editor__status-bar'
+                    style={{backgroundColor: editorState.statusBarColor}}
+                    onClick={() => {
```

Comment:
(Post-merge) Could you also pull these out into callback instead of defining them in the JSX? It's arguably okay to define callbacks inline in JSX, but we should try to keep that to when the logic inside them is simpler than this